### PR TITLE
Dynamic Extension Settings & User Overridden Person mapping

### PR DESCRIPTION
**Pull request by majora2007**
Refactored how settings work within the app. Now, there is a singleton `Settings` which can be used from anywhere. The class will load and save config from settings.ini (but can be overridden).  

Next, extensions (in this case, MetadataSources) have their own settings self contained. These settings will be inserted into Settings GUI at runtime and after saving, a callback on each extension will be called to update any internal state. 

This PR also includes some cleanup on how widgets are stored to ensure there is a consistent submodule they are loaded from, clear inheritance, and some QoL on the code. 

I've also updated so that we only import GUI code on the GUI flow. This should help with tkinker being required even for CLI invocation. 

Lastly, to showcase the new settings, external metadata now has custom overrides for how tags are translated to ComicInfo fields. 

TLDR:
- Bumped to Python 3.11
- Refactored all settings in the app to a singleton: Settings
- Refactored many GUI components into submodules and split into separate files (one class per file)
- Only when triggering GUI flow will tinker be imported
- MetadataSources now have their own settings and will now at runtime save to settings.ini and dynamically load in Settings Widget
- Lots of code cleanup throughout codebase

### CODE REVIEW

**Comment by majora2007** [comment]
Please hold on this, as discussed with Promidius, I will be adding ability for MetadataSource to have their own settings and be able to allow the user to provide custom mappings, rather than having a fixed set of mappings in the code.

**Comment by majora2007** [comment]
This is basically ready for review. I need to write Unit Tests for the new code and also handle the case where there is no settings.ini (only occurs for unit tests) and how to populate the default values.

File: MangaManager/Extensions/WebpConverter/WebpConverter.py [hunk 1]
```diff
@@ -10,12 +10,16 @@
 from tkinter import filedialog
 
 from Extensions.Interface import IExtensionApp
-from src import settings_class
+#from src import settings_class
```
Comment (ThePromidius) [inline MangaManager/Extensions/WebpConverter/WebpConverter.py:13]:
```suggestion

```

File: MangaManager/Extensions/WebpConverter/WebpConverter.py [hunk 2]
```diff
@@ -10,12 +10,16 @@
...
 from src.Common.loadedcomicinfo import LoadedComicInfo
 from src.Common.utils import ShowPathTreeAsDict
 from src.MetadataManager.GUI.widgets import ScrolledFrameWidget, ProgressBarWidget
+from src.Settings import default_settings
+from src.Settings.Settings import Settings
+
+#settings = settings_class.get_setting("WebpConverter")
+#settings = default_settings['WebpConverter']
```
Comment (ThePromidius) [inline MangaManager/Extensions/WebpConverter/WebpConverter.py:21]:
```suggestion
from src.Settings.Settings import Settings

```

File: MangaManager/src/Common/loadedcomicinfo.py [hunk 3]
```diff
@@ -30,7 +31,7 @@
 _LOG_TAG_WRITE_META = 'Write Meta'
 _LOG_TAG_RECOMPRESSING = "Recompressing"
 
-settings = settings_class.get_setting("main")
+# settings = default_settings.get_setting("main")
```
Comment (ThePromidius) [inline MangaManager/src/Common/loadedcomicinfo.py:34]:
```suggestion

```

File: MangaManager/src/Common/loadedcomicinfo.py [hunk 4]
```diff
@@ -13,10 +13,11 @@
 from PIL import Image, ImageTk
 from lxml.etree import XMLSyntaxError
 
-from src import settings_class as settings_class
+# from src import settings_class as settings_class
 from src.Common.errors import CorruptedComicInfo, BadZipFile
 from src.Common.utils import IS_IMAGE_PATTERN, convert_to_webp
 from src.Common.utils import obtain_cover_filename, get_new_webp_name
+# from src.Settings import default_settings
```
Comment (ThePromidius) [inline MangaManager/src/Common/loadedcomicinfo.py:20]:
```suggestion

```

File: MangaManager/src/Common/loadedcomicinfo.py [hunk 5]
```diff
@@ -13,10 +13,11 @@
 from PIL import Image, ImageTk
 from lxml.etree import XMLSyntaxError
 
-from src import settings_class as settings_class
+# from src import settings_class as settings_class
```
Comment (ThePromidius) [inline MangaManager/src/Common/loadedcomicinfo.py:16]:
```suggestion

```

File: MangaManager/src/MetadataManager/MetadataManagerGUI.py [hunk 6]
```diff
@@ -19,15 +22,14 @@
...
 from src.Common.loadedcomicinfo import LoadedComicInfo
-from src.MetadataManager.GUI.widgets import ComboBoxWidget, OptionMenuWidget, WidgetManager, ButtonWidget, ControlManager, HyperlinkLabel
-from src.MetadataManager.GUI.settings import SettingsWidgetManager
+from src.MetadataManager.GUI.widgets import ComboBoxWidget, OptionMenuWidget, WidgetManager, ButtonWidget, \
+    HyperlinkLabelWidget
+from src.MetadataManager.GUI.SettingsWidgetManager import SettingsWidgetManager
 from src.MetadataManager.MetadataManagerLib import MetadataManagerLib
 
-from src import settings_class
+# from src import settings_class
```
Comment (ThePromidius) [inline MangaManager/src/MetadataManager/MetadataManagerGUI.py:30]:
```suggestion

```

File: MangaManager/src/__init__.py [hunk 7]
```diff
@@ -24,12 +24,13 @@
...
-from src.settings import Settings
+#from src.settings import Settings
 
 CONFIG_PATH = Path(MM_PATH, _CONFIG_PATH)
 
-settings_class = Settings(CONFIG_PATH)
-settings_class.load_settings()
+
+# settings_class = Settings(CONFIG_PATH)
+# settings_class.load_settings()
```
Comment (ThePromidius) [inline MangaManager/src/__init__.py:33]:
```suggestion
```

File: MangaManager/src/MetadataManager/GUI/widgets/CanvasCoverWidget.py [hunk 8]
```diff
@@ -7,14 +7,14 @@
 from PIL import Image, ImageTk
 from pkg_resources import resource_filename
 
-from src import settings_class
+# from src import settings_class
```
Comment (ThePromidius) [inline MangaManager/src/MetadataManager/GUI/widgets/CanvasCoverWidget.py:10]:
```suggestion
```

File: MangaManager/src/MetadataManager/MetadataManagerLib.py [hunk 9]
```diff
@@ -209,15 +211,13 @@ def preview_export(self, loaded_cinfo):
         # loaded_cinfo.cinfo_object.export(export, 0)
         # print(export.getvalue())
```
Comment (ThePromidius) [inline MangaManager/src/MetadataManager/MetadataManagerLib.py:212]:
```suggestion
```

File: MangaManager/src/__init__.py [hunk 10]
```diff
@@ -24,12 +24,13 @@
 SOURCES_DIR = Path(sub_mm_path, "ExternalSources")
 SOURCES_DIR.mkdir(exist_ok=True)
 
-from src.settings import Settings
+#from src.settings import Settings
```
Comment (ThePromidius) [inline MangaManager/src/__init__.py:27]:
```suggestion
```

**Review by ThePromidius — COMMENTED** [review]
Remove dead and outdated commented code

File: MangaManager/src/MetadataManager/GUI/SettingsWidgetManager.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
```
Comment (ThePromidius) [inline MangaManager/src/MetadataManager/GUI/SettingsWidgetManager.py:1]: I had planned to add settings to the TLI.

I'm wondering whether we should tweak this to support CLI it or just make a new implementation for it.

**Comment by ThePromidius** [comment]
tried to run it and got this error:
```log
sub_mm_path:I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager
Found extensions: ['I:\\Mi unidad\\Programacion\\Python\\MangaManagerV2\\MangaManager\\Extensions\\WebpConverter\\WebpConverter.py', 'I:\\Mi unidad\\Programacion\\Python\\MangaManagerV2\\MangaManager\\Extensions\\CoverDownloader\\CoverDownloader.py']
Failed to Load extension I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\Extensions\WebpConverter\WebpConverter.py
Traceback (most recent call last):
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\DynamicLibController\extension_manager.py", line 55, in load_extensions
    extension_module = importlib.import_module(f"Extensions.{'.'.join(extract_folder_and_module(extension_file))}",package=EXTENSIONS_DIRECTORY)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\galla\AppData\Local\Programs\Python\Python311\Lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\Extensions\WebpConverter\WebpConverter.py", line 14, in <module>
    from src.Common.loadedcomicinfo import LoadedComicInfo
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\Common\loadedcomicinfo.py", line 512, in <module>
    from src.MetadataManager.comicinfo import ComicInfo, parseString
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\MetadataManager\__init__.py", line 7, in <module>
    from src.Layouts import layout_factory
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\Layouts\__init__.py", line 3, in <module>
    from src.Layouts.default_layout import Layout as DefLayout
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\Layouts\default_layout.py", line 14, in <module>
    from src.MetadataManager.MetadataManagerGUI import GUIApp
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\MetadataManager\MetadataManagerGUI.py", line 27, in <module>
    from src.MetadataManager.GUI.SettingsWidgetManager import SettingsWidgetManager
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\MetadataManager\GUI\SettingsWidgetManager.py", line 7, in <module>
    from main import providers
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\main.py", line 7, in <module>
    from ExternalSources.MetadataSources import ScraperFactory
ImportError: cannot import name 'ScraperFactory' from partially initialized module 'ExternalSources.MetadataSources' (most likely due to a circular import) (I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\ExternalSources\MetadataSources\__init__.py)
Exception loading the extensions
Traceback (most recent call last):
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\__init__.py", line 43, in <module>
    loaded_extensions = load_extensions(EXTENSIONS_DIR)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\DynamicLibController\extension_manager.py", line 65, in load_extensions
    for cls in extension_module.__dict__.values()
               ^^^^^^^^^^^^^^^^
UnboundLocalError: cannot access local variable 'extension_module' where it is not associated with a value
2023-02-05 14:04:20,004 -                 root -     INFO - INFO LEVEL - MAIN MODULE
2023-02-05 14:04:20,040 -                 root -     INFO - INFO LEVEL - MAIN MODULE
2023-02-05 14:04:20,387 -                 root -     INFO - Loading Config from: I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\settings.ini
2023-02-05 14:04:20,580 -                 root -     INFO - Loading Config from: I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\settings.ini
2023-02-05 14:04:20,583 -                 root -     INFO - Starting: GUI Manga Manager. Welcome
2023-02-05 14:04:20,584 -                 root -     INFO - Loading Config from: I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\settings.ini
Traceback (most recent call last):
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\main.py", line 94, in <module>
    execute_gui()
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\MetadataManager\__init__.py", line 17, in execute_gui
    if not Settings().get(SettingHeading.ExternalSources, 'default_metadata_source') in list([source.name for source in sources_factory.get("MetadataSources")]):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "I:\Mi unidad\Programacion\Python\MangaManagerV2\MangaManager\src\Settings\Settings.py", line 50, in get
    return self.config_parser[section][key].strip()
           ~~~~~~~~~~~~~~~~~~^^^^^^^^^
  File "C:\Users\galla\AppData\Local\Programs\Python\Python311\Lib\configparser.py", line 978, in __getitem__
    raise KeyError(key)
KeyError: <SettingHeading.ExternalSources: 'External Sources'>

Process finished with exit code 1
```

... circular imports

File: MangaManager/src/MetadataManager/MetadataManagerLib.py (hunk 9 above)
Comment (majora2007) [reply MangaManager/src/MetadataManager/MetadataManagerLib.py:212]: This isn't related to my changes, I didn't remove any dead code you already had present. Do you want me to proceed or leave this for another PR?

File: MangaManager/src/MetadataManager/GUI/SettingsWidgetManager.py (hunk 11 above)
Comment (majora2007) [reply MangaManager/src/MetadataManager/GUI/SettingsWidgetManager.py:1]: CLI can inherit the settings from extensions without any need to talk to SettingsWidgetManager btw.
